### PR TITLE
mixin proposal

### PR DESCRIPTION
**Pull request by condnsdmatters**

### CODE REVIEW

File: gpflow/models/training_interface.py [hunk 1]
```diff
@@ -12,25 +12,25 @@
...
 
 
-class InternalDataTrainingInterface:
+class InternalDataTrainingLossMixin:
     def training_loss(self):
         """
         Specify a training loss for this model
         """
-        raise NotImplementedError
+        return -self.maximum_a_posteriori_objective(data)
```
Comment (st--) [inline gpflow/models/training_interface.py]: I presume you mean
```suggestion
        return -self.maximum_a_posteriori_objective()
```
?

File: gpflow/models/mixins.py [hunk 2]
```diff
@@ -44,3 +44,11 @@ def closure():
             return training_loss(batch)
 
         return closure
+
+
+class MCMCTrainingLossMixin(InternalDataTrainingLossMixin):
```
Comment (joelberkeley-pio) [inline gpflow/models/mixins.py]: i don't know what MCMC is. clearly there's something about MCMC that means it has a different training loss. Should this also be conveyed in the names of the classes `InternalDataTrainingLossMixin` and the other one? Why, for example, don't they use the `log_posterior_density`? what's special about them?

File: gpflow/models/model.py [hunk 3]
```diff
@@ -46,41 +52,25 @@ def log_prior_density(self) -> tf.Tensor:
...
         """
-        return self._call_maximum_likelihood_objective(data) + self.log_prior_density()
+        return self.maximum_likelihood_objective(*args, **kwargs) + self.log_prior_density()
 
-    def training_loss(self, data: Optional[Data] = None) -> tf.Tensor:
+    def maximum_a_posteriori_objective(self, *args, **kwargs) -> tf.Tensor:
         """
         Minimization objective for TensorFlow optimizers (including
         gpflow.optimizers.Scipy). Includes the log prior density for maximum
         a-posteriori (MAP) estimation.
```
Comment (st--) [inline gpflow/models/model.py]:
```suggestion
        Maximum a-posteriori objective, assuming that maximum_likelihood_objective() is the log marginal likelihood.
```

File: gpflow/models/mixins.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+            batch = next(data) if isinstance(data, collections.abc.Iterator) else data
+            return training_loss(batch)
+
+        return closure
+
+
+class MCMCTrainingLossMixin(InternalDataTrainingLossMixin):
+    def training_loss(self):
+        """
+        Specify a training loss for this model
```
Comment (st--) [inline gpflow/models/mixins.py]:
```suggestion
        Training loss for gradient-based relaxation of MCMC models.
```

File: gpflow/models/mixins.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    def training_loss_closure(self, jit=True) -> Callable[[], tf.Tensor]:
+        if jit:
+            return tf.function(self.training_loss)
+        return self.training_loss
+
+
+class ExternalDataTrainingLossMixin:
+    def training_loss(self, data):
+        """
+        Specify a training loss for this model
```
Comment (st--) [inline gpflow/models/mixins.py]:
```suggestion
        Training loss for models that do not encapsulate the data.
        
        :param data: the data to be passed to the model objective.
```

File: gpflow/models/mixins.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
+import abc
```
Comment (st--) [inline gpflow/models/mixins.py]:
```suggestion
"""
This module defines training interfaces (realized as mix-in classes) that provide the minimization objective for TensorFlow optimizers (including gpflow.optimizers.Scipy). This `training_loss`  includes the log prior density for maximum a-posteriori (MAP) estimation.
"""

import abc
```

Comment (st--) [reply gpflow/models/mixins.py]: Going to rename this module to training_interface.py as it provides the data types etc.

File: doc/source/notebooks/tailor/mixture_density_network.pct.py [hunk 7]
```diff
@@ -85,7 +85,7 @@ def sinusoidal_data(N, noise):
 
 # %%
 import gpflow
-from gpflow.models import BayesianModel
+from gpflow.models import ExternalDataBayesianModel
```
Comment (st--) [inline doc/source/notebooks/tailor/mixture_density_network.pct.py]:
```suggestion
from gpflow.models import BayesianModel, ExternalDataTrainingLossMixin
```

File: doc/source/notebooks/tailor/mixture_density_network.pct.py [hunk 8]
```diff
@@ -97,7 +97,7 @@ def sinusoidal_data(N, noise):
 # %%
 from typing import Callable, Optional, Tuple
 
-class MDN(BayesianModel):
+class MDN(ExternalDataBayesianModel):
```
Comment (st--) [inline doc/source/notebooks/tailor/mixture_density_network.pct.py]:
```suggestion
class MDN(BayesianModel, ExternalDataTrainingLossMixin):
```

File: gpflow/models/__init__.py [hunk 9]
```diff
@@ -18,6 +18,11 @@
 from .gpmc import GPMC
 from .gpr import GPR
 from .model import BayesianModel, GPModel
+from .mixins import (
```
Comment (st--) [inline gpflow/models/__init__.py]:
```suggestion
from .training_interface import (
```

File: gpflow/models/mixins.py [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+InputData = tf.Tensor
+OutputData = tf.Tensor
+RegressionData = Tuple[InputData, OutputData]
+Data = TypeVar("Data", RegressionData, InputData)
+
+
+class InternalDataTrainingLossMixin:
+    def training_loss(self):
+        """
+        Specify a training loss for this model
```
Comment (st--) [inline gpflow/models/mixins.py]:
```suggestion
        Training loss for models that encapsulate their data.
```

File: gpflow/models/__init__.py (hunk 9 above)
Comment (st--) [reply gpflow/models/__init__.py]: (as it defines all the types I think "mixin" is rather confusing; it does all relate to 'data' so "training_interface" seems appropriate)

File: gpflow/models/model.py [hunk 11]
```diff
@@ -18,6 +18,16 @@
 
 import numpy as np
 import tensorflow as tf
+from tensorflow.python.data.ops import iterator_ops
+
+from .training_interface import (
+    InternalDataTrainingLossMixin,
+    ExternalDataTrainingLossMixin,
```
Comment (st--) [inline gpflow/models/model.py]: I think these aren't actually needed here ?
```suggestion
```
Or if you want them in here to expose them to people who want to build models, we should add the mcmc one:
```suggestion
    InternalDataTrainingLossMixin,
    ExternalDataTrainingLossMixin,
    MCMCTrainingLossMixin,
```

File: gpflow/models/model.py [hunk 12]
```diff
@@ -18,6 +18,16 @@
 
 import numpy as np
 import tensorflow as tf
+from tensorflow.python.data.ops import iterator_ops
```
Comment (st--) [inline gpflow/models/model.py]:
```suggestion
```

File: gpflow/models/mixins.py [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+from typing import Callable, Optional, Tuple, TypeVar
+
+
+import tensorflow as tf
+import numpy as np
+
+InputData = tf.Tensor
+OutputData = tf.Tensor
+RegressionData = Tuple[InputData, OutputData]
+Data = TypeVar("Data", RegressionData, InputData)
```
Comment (joelberkeley-pio) [inline gpflow/models/mixins.py]:
```suggestion
Data = Union[RegressionData, InputData]
```

File: gpflow/models/mixins.py [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+
+    def training_loss_closure(self, data: Data, jit=True) -> Callable[[], tf.Tensor]:
+        training_loss = self.training_loss
+        if jit:
+            training_loss = tf.function(
+                self.training_loss
+            )  # TODO need to add correct input_signature here to allow for differently sized minibatches
+
+        def closure():
+            batch = next(data) if isinstance(data, collections.abc.Iterator) else data
```
Comment (joelberkeley-pio) [inline gpflow/models/mixins.py]: when is a `tf.Tensor` an `Iterator`?

Comment (joelberkeley-pio) [inline gpflow/models/mixins.py]: what happens when `next` raises `StopIteration`? Will the optimizer handle that?

File: gpflow/models/mixins.py [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+        """
+        Specify a training loss for this model
+        """
+        return -self.maximum_a_posteriori_objective(data)
+
+    def training_loss_closure(self, data: Data, jit=True) -> Callable[[], tf.Tensor]:
+        training_loss = self.training_loss
+        if jit:
+            training_loss = tf.function(
+                self.training_loss
```
Comment (joelberkeley-pio) [inline gpflow/models/mixins.py]:
```suggestion
               training_loss
```

File: gpflow/models/__init__.py (hunk 9 above)
Comment (condnsdmatters) [reply gpflow/models/__init__.py]: i think mixins keeps it clearer that the pattern we are going for is a mixin pattern

File: gpflow/models/mixins.py (hunk 14 above)
Comment (condnsdmatters) [reply gpflow/models/mixins.py]: good question: i dont know. normally you would use a tf.data.Datatset and set it to loop so you never hit `StopIteration`

Comment (joelberkeley-pio) [reply gpflow/models/mixins.py]: i think we'd need to be clear on that before we can be confident we support batch processing

File: gpflow/models/mixins.py (hunk 13 above)
Comment (st--) [reply gpflow/models/mixins.py]: what's the difference ?

File: gpflow/models/mixins.py (hunk 2 above)
Comment (st--) [reply gpflow/models/mixins.py]: log_posterior_density is the right thing to compute for an MCMC model. maximum likelihood / map objective is the right thing to compute for the other models.

File: gpflow/models/mixins.py (hunk 6 above)
Comment (st--) [reply gpflow/models/mixins.py]: agreed with @condnsdmatters to move types into `gpflow.base` and rename this to `gpflow.models.training_mixins`

File: gpflow/models/mixins.py (hunk 14 above)
Comment (st--) [reply gpflow/models/mixins.py]: a tensor isn't an iterator, it's an iterator if it's a dataset to iterate over (e.g. minibatching)

Comment (st--) [reply gpflow/models/mixins.py]: you need to call `.repeat()` on your dataset (otherwise you can't train properly!) and then it loops forever

File: gpflow/models/mixins.py [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+class ExternalDataTrainingLossMixin:
+    def training_loss(self, data):
+        """
+        Training loss for models that do not encapsulate the data.
+        
+        :param data: the data to be used for computing the model objective.
+        """
+        return -self.maximum_a_posteriori_objective(data)
+
+    def training_loss_closure(self, data: Data, jit=True) -> Callable[[], tf.Tensor]:
```
Comment (st--) [inline gpflow/models/mixins.py]:
```suggestion
    def training_loss_closure(self, data: Union[Data, collections.abc.Iterator], jit=True) -> Callable[[], tf.Tensor]:
```

File: gpflow/models/mixins.py (hunk 14 above)
Comment (st--) [reply gpflow/models/mixins.py]: see type signature amendment above

Comment (st--) [reply gpflow/models/mixins.py]: I think it is going to be fine as it is for now. If you exhaust your training set but insist on keep having more optimisation steps your code crashes :woman_shrugging:

Comment (joelberkeley-pio) [reply gpflow/models/mixins.py]: how easy is that to do? isn't it just a case of specifying too high a number of iterations? that sounds horrendously fragile

File: gpflow/models/mixins.py (hunk 2 above)
Comment (joelberkeley-pio) [reply gpflow/models/mixins.py]: what i wonder is if we can be a bit more be a bit more descriptive in the names of the other mixins
